### PR TITLE
Adding MKL matmul op

### PR DESCRIPTION
**Pull request by vivek-rane**
Adding build, op and registration for the MKL matmul op.

### CODE REVIEW

File: tensorflow/core/kernels/mkl_matmul_op.cc [hunk 1]
```diff
@@ -0,0 +1,157 @@
+/* Copyright 2015 The TensorFlow Authors. All Rights Reserved.
```
Comment (rmlarsen) [inline tensorflow/core/kernels/mkl_matmul_op.cc]: This is great. Can you also add version of BatchMatmul based on MKL. Many matmul calls end up in BatchMatmul if the dimensions are not known at graph construction time. See: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/math_ops.py#L1818

Comment (andydavis1) [inline tensorflow/core/kernels/mkl_matmul_op.cc]: +1 to Rasmus' request...

File: tensorflow/core/kernels/mkl_matmul_op.cc [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+#include "tensorflow/core/kernels/fill_functor.h"
+#include "third_party/mkl/include/mkl_cblas.h"
+
+namespace tensorflow {
+
+typedef Eigen::ThreadPoolDevice CPUDevice;
+
+template <typename Device, typename T, bool USE_CUBLAS>
+class MatMulOpMKL : public OpKernel {
+  void MklBlasGemm(bool transa, bool transb, const int m, const int n,
```
Comment (andydavis1) [inline tensorflow/core/kernels/mkl_matmul_op.cc]: Can all of these "MklBlasGemm" functions be moved to after the "private:" currently on line 138?

File: tensorflow/core/kernels/mkl_matmul_op.cc [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+#include "tensorflow/core/framework/op.h"
+#include "tensorflow/core/framework/op_kernel.h"
+#include "tensorflow/core/framework/register_types.h"
+#include "tensorflow/core/kernels/fill_functor.h"
+#include "third_party/mkl/include/mkl_cblas.h"
+
+namespace tensorflow {
+
+typedef Eigen::ThreadPoolDevice CPUDevice;
+
```
Comment (andydavis1) [inline tensorflow/core/kernels/mkl_matmul_op.cc]: Can you add some high-level documentation at the top of this file?

File: tensorflow/core/kernels/mkl_matmul_op.cc [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+                   const int k, const float* a, const int lda, const float* b,
+                   const int ldb, float* c, const int ldc) {
+    const float alpha = 1.0f;
+    const float beta = 0.0f;
+    cblas_sgemm(CblasRowMajor, transa ? CblasTrans : CblasNoTrans,
+                transb ? CblasTrans : CblasNoTrans, m, n, k, alpha, a, lda, b,
+                ldb, beta, c, ldc);
+  }
+
+  void MklBlasGemm(bool transa, bool transb, const int m, const int n,
```
Comment (andydavis1) [inline tensorflow/core/kernels/mkl_matmul_op.cc]: Each of these functions could have short document documentation.

Comment (andydavis1) [inline tensorflow/core/kernels/mkl_matmul_op.cc]: For example, explain the function arguments: "lda", "b", "ldb", etc..

File: tensorflow/core/kernels/mkl_matmul_op.cc [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+                   const int k, const double* a, const int lda, const double* b,
+                   const int ldb, double* c, const int ldc) {
+    const double alpha = 1.0;
+    const double beta = 0.0;
+    cblas_dgemm(CblasRowMajor, transa ? CblasTrans : CblasNoTrans,
+                transb ? CblasTrans : CblasNoTrans, m, n, k, alpha, a, lda, b,
+                ldb, beta, c, ldc);
+  }
+
+  void MklBlasGemm(bool transa, bool transb, const int m, const int n,
```
Comment (andydavis1) [inline tensorflow/core/kernels/mkl_matmul_op.cc]: It looks like most of these functions only differ from each other by data type (each of which call different cblas_XXX variants).

Is it possible to templatize on data type here "typename T" and have template specializations for each of: float, double, std::complex<float> , and std::complex<double>?  

 Then your MklBlasGemm call on line 134, would turn into  MklBlasGemm<T>(...)

File: tensorflow/core/kernels/mkl_matmul_op.cc (hunk 1 above)
Comment (vamsi-sripathi) [inline tensorflow/core/kernels/mkl_matmul_op.cc]: Thanks for the feedback. I haven't yet looked at batch Matmul, but, yes, we will add MKL support there too

File: tensorflow/core/kernels/mkl_matmul_op.cc (hunk 3 above)
Comment (vamsi-sripathi) [inline tensorflow/core/kernels/mkl_matmul_op.cc]: Done.

File: tensorflow/core/kernels/mkl_matmul_op.cc (hunk 4 above)
Comment (vamsi-sripathi) [inline tensorflow/core/kernels/mkl_matmul_op.cc]: I added documentation for the function parameters.

File: tensorflow/core/kernels/mkl_matmul_op.cc (hunk 2 above)
Comment (vamsi-sripathi) [inline tensorflow/core/kernels/mkl_matmul_op.cc]: I modified the MklBlasGemm to be private member functions

File: tensorflow/core/kernels/mkl_matmul_op.cc (hunk 5 above)
Comment (vamsi-sripathi) [inline tensorflow/core/kernels/mkl_matmul_op.cc]: As discussed previously, I haven't updated the current implementation

**Review by andydavis1 — APPROVED** [review]

**Comment by andydavis1** [comment]
Looks good. As discussed offline, rmlarsen's comment will be addressed in a separate PR.

**Comment by andydavis1** [comment]
@tensorflow-jenkins test this please

**Comment by vrv** [comment]
Please fix the buildifier error mentioned in https://ci.tensorflow.org/job/tensorflow-pull-requests-sanity/3098/console and we can retest.  Thanks!

**Comment by vamsi-sripathi** [comment]
Ran buildifier on tensorflow/core/BUILD, tensorflow/core/kernels/BUILD and pushed the changes

**Comment by vrv** [comment]
@tensorflow-jenkins test this please

**Comment by vrv** [comment]
@tensorflow-jenkins test this please

**Comment by vrv** [comment]
Weird, I'm seeing the following now:

```
ERROR: missing input file '//third_party/mkl:libmklml_intel.so'.
ERROR: /var/lib/jenkins/workspace/tensorflow-pull-requests-cpu-python3/bazel-ci_build-cache/.cache/bazel/_bazel_jenkins/eab0d61a99b6696edb3d2aff87b585e8/external/nasm/BUILD:8:1: C++ compilation of rule '@nasm//:nasm' failed: gcc failed: error executing command 
  (cd /var/lib/jenkins/workspace/tensorflow-pull-requests-cpu-python3/bazel-ci_build-cache/.cache/bazel/_bazel_jenkins/eab0d61a99b6696edb3d2aff87b585e8/execroot/workspace && \
  exec env - \
    LD_LIBRARY_PATH='' \
    PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin \
  /usr/bin/gcc -U_FORTIFY_SOURCE '-D_FORTIFY_SOURCE=1' -fstack-protector -Wall -Wl,-z,-relro,-z,now -B/usr/bin -B/usr/bin -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-canonical-system-headers -fno-omit-frame-pointer -g0 -O2 -DNDEBUG -ffunction-sections -fdata-sections -MD -MF bazel-out/local-py3-opt/bin/external/nasm/_objs/nasm/external/nasm/preproc.d -iquote external/nasm -iquote bazel-out/local-py3-opt/genfiles/external/nasm -iquote external/bazel_tools -iquote bazel-out/local-py3-opt/genfiles/external/bazel_tools -isystem external/bazel_tools/tools/cpp/gcc3 -w '-std=c99' -fno-canonical-system-headers -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c external/nasm/preproc.c -o bazel-out/local-py3-opt/bin/external/nasm/_objs/nasm/external/nasm/preproc.o): com.google.devtools.build.lib.shell.AbnormalTerminationException: Process terminated by signal 15.
ERROR: /workspace/third_party/mkl/BUILD:11:1: //third_party/mkl:intel_binary_blob: missing input file '//third_party/mkl:libmklml_intel.so'.
```

**Comment by mahmoud-abuzaina** [comment]
I think this error should appear if you build with --config=mkl while not enabling mkl when you configure (currently it is disabled in the configure file). Can you confirm if that is the case?

**Comment by andydavis1** [comment]
Where does this PR stand currently? @vrv  Can you describe what configure and build settings you used when you got the error you mentioned 5 days back?

**Comment by vrv** [comment]
I'm just pointing at the test results... click on "details' and then 'console log" for the relevant failing tests.

**Comment by vamsi-sripathi** [comment]
Previously, we were not testing with "bazel test" in our setup so we could not reproduce the error earlier. Now, I updated the third_party/mkl/BUILD to fix the test failures due to missing MKL libs

**Comment by andydavis1** [comment]
@tensorflow-jenkins test this please

**Comment by maciekcc** [comment]
Jenkins, test this please.

**Pull request merged**